File: PluginsAndFeatures/azure-toolkit-for-rider/ReSharper.Azure/src/Azure.Daemon/FunctionApp/FunctionAppDaemonHost.cs
// Copyright (c) 2020-2022 JetBrains s.r.o.
//
// All rights reserved.
//
// MIT License
//
// Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated
// documentation files (the "Software"), to deal in the Software without restriction, including without limitation
// the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and
// to permit persons to whom the Software is furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all copies or substantial portions of
// the Software.
//
// THE SOFTWARE IS PROVIDED *AS IS*, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO
// THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
// TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

using System.Linq;
using JetBrains.Lifetimes;
using JetBrains.ProjectModel;
using JetBrains.ProjectModel.MSBuild;
using JetBrains.ProjectModel.Properties;
using JetBrains.Rd.Tasks;
using JetBrains.RdBackend.Common.Features;
using JetBrains.ReSharper.Resources.Shell;
using JetBrains.Rider.Azure.Model;
using JetBrains.Util;

namespace JetBrains.ReSharper.Azure.Daemon.FunctionApp
{
    [SolutionComponent]
    public class FunctionAppDaemonHost
    {
        private readonly ISolution _solution;
        private readonly FunctionAppDaemonModel _model;

        public FunctionAppDaemonHost(ISolution solution)
        {
            _solution = solution;
            
            _model = solution.GetProtocolSolution().GetFunctionAppDaemonModel();
            _model.GetAzureFunctionsVersion.Set(GetAzureFunctionsVersionHandler);
        }
        
        public void RunFunctionApp(string methodName, string functionName, string projectFilePath)
        {
            _model.RunFunctionApp(new FunctionAppRequest(methodName, functionName, projectFilePath));
        }

        public void DebugFunctionApp(string methodName, string functionName, string projectFilePath)
        {
            _model.DebugFunctionApp(new FunctionAppRequest(methodName, functionName, projectFilePath));
        }

        public void TriggerFunctionApp(string methodName, string functionName, string projectFilePath)
        {
            _model.TriggerFunctionApp(new FunctionAppRequest(methodName, functionName, projectFilePath));
        }
        
        private RdTask<string> GetAzureFunctionsVersionHandler(Lifetime lifetime, AzureFunctionsVersionRequest request)
        {
            using (ReadLockCookie.Create())
            {
                var project = _solution.FindProjectByProjectFilePath(VirtualFileSystemPath.Parse(request.ProjectFilePath, InteractionContext.SolutionContext));

                var azureFunctionsVersion = project?.GetRequestedProjectProperties(MSBuildProjectUtil.AzureFunctionsVersionProperty).FirstOrDefault();

                return RdTask<string>.Successful(azureFunctionsVersion);
            }
        }
    }
}

File: PluginsAndFeatures/azure-toolkit-for-rider/src/main/kotlin/org/jetbrains/plugins/azure/functions/daemon/FunctionAppDaemonHost.kt
///**
// * Copyright (c) 2020-2022 JetBrains s.r.o.
// *
// * All rights reserved.
// *
// * MIT License
// *
// * Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated
// * documentation files (the "Software"), to deal in the Software without restriction, including without limitation
// * the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and
// * to permit persons to whom the Software is furnished to do so, subject to the following conditions:
// *
// * The above copyright notice and this permission notice shall be included in all copies or substantial portions of
// * the Software.
// *
// * THE SOFTWARE IS PROVIDED *AS IS*, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO
// * THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
// * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// * SOFTWARE.
// */
//
//package org.jetbrains.plugins.azure.functions.daemon
//
//import com.intellij.execution.Executor
//import com.intellij.execution.ProgramRunnerUtil
//import com.intellij.execution.RunManager
//import com.intellij.execution.RunnerAndConfigurationSettings
//import com.intellij.execution.configurations.ConfigurationTypeUtil
//import com.intellij.execution.executors.DefaultDebugExecutor
//import com.intellij.execution.executors.DefaultRunExecutor
//import com.intellij.openapi.actionSystem.ActionPlaces
//import com.intellij.openapi.actionSystem.ex.ActionUtil
//import com.intellij.openapi.actionSystem.impl.SimpleDataContext
//import com.intellij.openapi.project.Project
//import com.intellij.util.execution.ParametersListUtil
//import com.jetbrains.rd.ui.bedsl.extensions.valueOrEmpty
//import com.jetbrains.rdclient.util.idea.LifetimedProjectComponent
//import com.jetbrains.rider.azure.model.functionAppDaemonModel
//import com.jetbrains.rider.model.RunnableProject
//import com.jetbrains.rider.model.runnableProjectsModel
//import com.jetbrains.rider.projectView.solution
//import org.jetbrains.plugins.azure.functions.actions.TriggerAzureFunctionAction
//import org.jetbrains.plugins.azure.functions.run.AzureFunctionsHostConfiguration
//import org.jetbrains.plugins.azure.functions.run.AzureFunctionsHostConfigurationFactory
//import org.jetbrains.plugins.azure.functions.run.AzureFunctionsHostConfigurationType
//
//@Suppress("ComponentNotRegistered")
//class FunctionAppDaemonHost(project: Project) : LifetimedProjectComponent(project) {
//
//    val model = project.solution.functionAppDaemonModel
//
//    init {
//        initRunFunctionAppHandler()
//        initDebugFunctionAppHandler()
//        initTriggerFunctionApp()
//    }
//
//    private fun initRunFunctionAppHandler() {
//        model.runFunctionApp.advise(componentLifetime) { functionAppRequest ->
//            runConfiguration(
//                    functionName = functionAppRequest.functionName,
//                    runnableProject = getRunnableProject(project, functionAppRequest.projectFilePath),
//                    executor = DefaultRunExecutor.getRunExecutorInstance())
//        }
//    }
//
//    private fun initDebugFunctionAppHandler() {
//        model.debugFunctionApp.advise(componentLifetime) { functionAppRequest ->
//            runConfiguration(
//                    functionName = functionAppRequest.functionName,
//                    runnableProject = getRunnableProject(project, functionAppRequest.projectFilePath),
//                    executor = DefaultDebugExecutor.getDebugExecutorInstance())
//        }
//    }
//
//    private fun initTriggerFunctionApp() {
//        model.triggerFunctionApp.advise(componentLifetime) { triggerFunctionRequest ->
//            val triggerAction = TriggerAzureFunctionAction(functionName = triggerFunctionRequest.functionName)
//
//            ActionUtil.invokeAction(
//                    triggerAction,
//                    SimpleDataContext.getProjectContext(project),
//                    ActionPlaces.EDITOR_GUTTER_POPUP,
//                    null,
//                    null)
//        }
//    }
//
//    private fun runConfiguration(
//            functionName: String,
//            runnableProject: RunnableProject,
//            executor: Executor
//    ) {
//        val runManager = RunManager.getInstance(project)
//
//        val existingSettings = findExistingConfigurationSettings(functionName, runnableProject.projectFilePath)
//
//        // Take existing configuration or generate a new one if configuration is missing.
//        val settings = existingSettings ?: let {
//            val configuration = createFunctionAppRunConfiguration(project, functionName, runnableProject)
//            val newSettings = runManager.createConfiguration(configuration, configuration.factory as AzureFunctionsHostConfigurationFactory)
//
//            runManager.setTemporaryConfiguration(newSettings)
//            runManager.addConfiguration(newSettings)
//
//            newSettings
//        }
//
//        runManager.selectedConfiguration = settings
//        ProgramRunnerUtil.executeConfiguration(settings, executor)
//    }
//
//    private fun findExistingConfigurationSettings(
//            functionName: String,
//            projectFilePath: String
//    ): RunnerAndConfigurationSettings? {
//        val runManager = RunManager.getInstance(project)
//
//        val configurationType = ConfigurationTypeUtil.findConfigurationType(AzureFunctionsHostConfigurationType::class.java)
//        val runConfigurations = runManager.getConfigurationsList(configurationType)
//
//        return runConfigurations.filterIsInstance<AzureFunctionsHostConfiguration>().firstOrNull { configuration ->
//            configuration.parameters.functionNames == functionName &&
//                    configuration.parameters.projectFilePath == projectFilePath
//        }?.let { configuration ->
//            runManager.findSettings(configuration)
//        }
//    }
//
//    private fun createFunctionAppRunConfiguration(
//            project: Project,
//            functionName: String,
//            runnableProject: RunnableProject
//    ): AzureFunctionsHostConfiguration {
//
//        val runManager = RunManager.getInstance(project)
//        val configurationType = ConfigurationTypeUtil.findConfigurationType(AzureFunctionsHostConfigurationType::class.java)
//
//        val factory = configurationType.factory
//        val template = runManager.getConfigurationTemplate(factory)
//
//        val configuration = template.configuration as AzureFunctionsHostConfiguration
//        patchConfigurationParameters(configuration, runnableProject, functionName)
//
//        return configuration
//    }
//
//    private fun patchConfigurationParameters(
//            configuration: AzureFunctionsHostConfiguration,
//            runnableProject: RunnableProject,
//            functionName: String
//    ) {
//        val projectOutput = runnableProject.projectOutputs.singleOrNull()
//
//        configuration.name = "$functionName (${runnableProject.fullName})"
//        configuration.parameters.functionNames = functionName
//
//        configuration.parameters.projectFilePath = runnableProject.projectFilePath
//        configuration.parameters.projectKind = runnableProject.kind
//        configuration.parameters.projectTfm = projectOutput?.tfm?.presentableName ?: ""
//        configuration.parameters.exePath = projectOutput?.exePath ?: ""
//        configuration.parameters.programParameters = ParametersListUtil.join(projectOutput?.defaultArguments
//                ?: listOf())
//        configuration.parameters.workingDirectory = projectOutput?.workingDirectory ?: ""
//    }
//
//    private fun getRunnableProject(
//            project: Project,
//            expectedProjectPath: String
//    ): RunnableProject {
//        val runnableProjects = project.solution.runnableProjectsModel.projects.valueOrEmpty()
//        return runnableProjects.find { runnableProject ->
//            runnableProject.projectFilePath == expectedProjectPath && runnableProject.kind == AzureRunnableProjectKinds.AzureFunctions }
//                ?: throw IllegalStateException(
//                        "Unable to find a project to run with path: '$expectedProjectPath', available project paths: " +
//                                runnableProjects.joinToString(", ", "'", "'") { it.projectFilePath})
//    }
//}